### PR TITLE
Sync sync supply points

### PR DESCRIPTION
**Pull request by esoergel**
Currently supply-point cases are updated via a signal that gets called only when you edit a location via the UI.  That's two layers of indirection, so here I call it directly in `location.save()`.  The first commit is the actual change, and the second is fixing all the tests it broke.

The test failures were due to two things:
- a bunch of tests were manually creating supply point cases, because they weren't created automatically
- Now that supply-point cases are being created automatically, you can't create a location without a real domain.  This seemed like a reasonable constraint, so I fixed the tests, rather than work around it to make the tests simpler.

@czue @benrudolph

### CODE REVIEW

**Comment by esoergel** [comment]
I'd definitely wait on tests, I almost certainly broke some.

File: corehq/apps/commtrack/tests/test_locations.py [hunk 1]
```diff
@@ -172,10 +169,10 @@ def test_can_set_locations(self):
         user = self.user
 
         loc1 = make_loc('secondloc')
-        sp1 = make_supply_point(self.domain.name, loc1)
+        sp1 = loc.linked_supply_point()
```
Comment (dimagimon) [inline corehq/apps/commtrack/tests/test_locations.py]: F821 undefined name 'loc'

File: corehq/apps/commtrack/tests/test_locations.py [hunk 2]
```diff
@@ -172,10 +169,10 @@ def test_can_set_locations(self):
...
 
         loc2 = make_loc('thirdloc')
-        sp2 = make_supply_point(self.domain.name, loc2)
+        sp2 = loc.linked_supply_point()
```
Comment (dimagimon) [inline corehq/apps/commtrack/tests/test_locations.py]: F821 undefined name 'loc'

File: corehq/apps/commtrack/models.py [hunk 3]
```diff
@@ -551,38 +546,19 @@ def get_data(self, value):
...
-            sql_loc.save()
-        except SQLLocation.DoesNotExist:
-            pass
+    supply_point = loc.linked_supply_point()
+    if supply_point:
+        supply_point.update_from_location(loc)
+        updated_supply_point = supply_point
+    else:
+        updated_supply_point = SupplyInterface.create_from_location(loc.domain, loc)
+    return updated_supply_point._id
```
Comment (czue) [inline corehq/apps/commtrack/models.py]: nice this is quite a bit simpler

**Comment by czue** [comment]
so I think one potentially significant downside of this approach is the potential performance hit. it looks like syncing the supply point will involve a domain lookup, a few couch calls to check if the supply point exists, and then submitting a form (which is many more couch calls and db writes). for clicking "save" on the location page that's not great but the user might expect it to be slow. for something like bulk importing locations that's going to be pretty painful (i don't actually know how that works currently).

there look like there are a lot of great changes in this PR that are just code cleanup that we should get in - i wonder if the main change should wait till we can do a more thorough audit of all the places that are impacted by that change

**Comment by czue** [comment]
(edited the above a bit for clarity)

File: corehq/apps/commtrack/models.py (hunk 3 above)
Comment (snopoke) [inline corehq/apps/commtrack/models.py]: :100:

**Comment by esoergel** [comment]
@czue I am also a bit worried about the performance implications, maybe we should throw this on staging and play with it there.  I do have some thoughts on the specifics:

I've been thinking of `Domain.get_by_name` as basically free - it's quickcached, and is pretty much guaranteed to have been called already recently.

As part of the other changes I'm making (in a branch not yet in PR), I'm moving `location.linked_supply_point()` to be basically:
```python
if not self.supply_point_id:
    return None
try:
    SupplyPointCase.get(self.supply_point_id)
except ResourceNotFound:
    return None
```
Which should be both faster, and a noop for locations where it's not relevant.

For locations which do have supply_points, the supply points need to be updated when the location is changed.  I could probably be more clever here about skipping the sync if no relevant fields were changed, but I do think that in general some variation of this is needed to keep supply point cases in sync with their locations.

The work I'm doing here is rapidly getting beyond the scope of what I intended originally when I started working on this feature, but I do think the changes are valuable.  What are your thoughts on prioritizing this versus propagating location type changes through a quicker, less elegant way?

**Comment by czue** [comment]
@esoergel just ping me again when you want me to look at this. as per our conversation i was signed off as of last week, but looks like you're still working on it.

**Comment by esoergel** [comment]
@czue will do, thanks.  I've got the other part of this feature finished and I want to throw it up on staging first.

**Comment by esoergel** [comment]
Closing to merge https://github.com/dimagi/commcare-hq/pull/9946 directly instead

**Pull request closed without merging**